File: src/games/postflop_nlhe.py
"""
Heads-Up No-Limit Hold'em — Preflop through River.

Designed for Deep CFR (not tabular CFR). Chance nodes are sampled
during MCCFR traversal rather than pre-expanded, since the game
tree is too large to enumerate (~10^14 information sets).

History encoding:
    (p0_cards, p1_cards, board, action1, action2, ...)

    p0_cards: tuple of 2 card ints (0-51)
    p1_cards: tuple of 2 card ints (0-51)
    board:    tuple of 0-5 card ints (grows as streets are dealt)
    actions:  'f' fold, 'c' check, 'k' call, 'r' raise, 'a' all-in

Streets determined by board length:
    0 cards = preflop
    3 cards = flop
    4 cards = turn
    5 cards = river

Card encoding: card = rank * 4 + suit
    rank: 0=2, 1=3, ..., 12=A
    suit: 0=♣, 1=♦, 2=♥, 3=♠

IMPLEMENTATION NOTE — state machine parity with the C++ engine
--------------------------------------------------------------
_parse_state is a deterministic replay of the SAME state machine implemented
in C++ (NLHEGame::apply_action in nlhe_game.cpp). This is required so that the
Python game tree is bit-identical to the C++ tree used to train the blueprints.
The key behaviours that distinguish this from a naive "betting round" model:

  * Position: SB (player 0) acts first preflop; BB (player 1) acts first
    postflop (the button/SB is in position).
  * CALL always ends the street immediately (advance to next street), including
    a preflop SB call — the BB does NOT get a subsequent option. This matches
    the C++ engine (it is non-standard poker, but it is what the engine does,
    and the engine defines the trained game).
  * CHECK uses last_aggressor: the first check of a street passes action to the
    opponent (last_aggressor -1 -> -2); the second check ends the street.
  * Raise sizing is the standard pot-sized raise on the pot AFTER the call:
    raise_add = (pot + owe) * frac, then total = owe + raise_add, capped at the
    stack. bet_amount mirrors NLHEGame::bet_amount.
"""

import numpy as np
from dataclasses import dataclass
from ..games.base import ExtensiveFormGame, History, Action, InfoSetKey
from ..abstraction.equity import evaluate_7card


STREET_NAMES = {0: "preflop", 3: "flop", 4: "turn", 5: "river"}

# Action char <-> meaning. Same external alphabet as before:
#   'f' fold, 'c' check, 'k' call, 'r' raise, 'a' all-in
# Internally 'f'/'c' both map to the C++ FOLD_OR_CHECK action (slot 0); which
# one it is depends on whether there is an outstanding bet (owe > 0).


@dataclass
class PostflopNLHE(ExtensiveFormGame):
    """
    Full HU NLHE for Deep CFR. Tree-identical to the C++ NLHEGame engine.

    Not compatible with tabular CFR (game tree too large).
    Use DeepCFRSolver which samples chance nodes.
    """
    starting_stack: float = 200.0  # In chips (BB = 2, SB = 1)
    max_raises_per_street: int = 2  # matches C++ NLHEGameConfig.max_raises
    sb: float = 1.0
    bb: float = 2.0
    # Raise sizes as fraction of pot (pot AFTER the call)
    raise_fractions: tuple[float, ...] = (0.75,)

    def num_players(self) -> int:
        return 2

    def initial_histories(self):
        """Not used by Deep CFR. Raises error for tabular solvers."""
        raise NotImplementedError(
            "PostflopNLHE is too large for tabular CFR. "
            "Use DeepCFRSolver with sample_deal()."
        )

    def sample_deal(self, rng: np.random.Generator) -> History:
        """Sample a random initial deal (hole cards + full board)."""
        cards = rng.choice(52, size=9, replace=False)
        p0 = (int(cards[0]), int(cards[1]))
        p1 = (int(cards[2]), int(cards[3]))
        full_board = tuple(int(c) for c in cards[4:9])
        return (p0, p1, full_board)

    def _get_actions(self, history: History) -> tuple:
        return history[3:]

    # ── Core state machine (mirrors C++ NLHEGame::apply_action) ──────────────
    def _parse_state(self, history: History) -> dict:
        """
        Deterministic replay of the C++ state machine. Returns the live state
        AFTER applying every action in the history. The keys returned match the
        previous public contract (stacks, invested, pot, to_call, street_idx,
        street_name, raises_this_street, current_player, folded, all_in).

        'invested' is reported as TOTAL invested across the hand
        (starting_stack - stack), so payoffs and the encoder see cumulative
        contributions, while 'street_invest' drives call sizing within a street.
        """
        actions = self._get_actions(history)

        stacks = [self.starting_stack - self.sb, self.starting_stack - self.bb]
        street_invest = [self.sb, self.bb]
        pot = self.sb + self.bb
        street = 0
        current_player = 0      # SB acts first preflop
        raises_this_street = 0
        last_aggressor = -1     # -1 = no bet yet this street, -2 = one check done
        folded = [False, False]
        all_in = False
        terminal = False

        def advance_street():
            nonlocal street, raises_this_street, last_aggressor, street_invest
            nonlocal current_player, terminal
            street += 1
            raises_this_street = 0
            last_aggressor = -1
            street_invest = [0.0, 0.0]
            if street >= 4:
                terminal = True          # showdown
            else:
                current_player = 1        # BB acts first postflop

        def bet_amount():
            p = current_player
            owe = street_invest[1 - p] - street_invest[p]
            effective_pot = pot + owe     # pot after calling
            raise_add = effective_pot * self.raise_fractions[0]
            return min(raise_add, stacks[p] - owe)

        for a in actions:
            if terminal:
                break
            p = current_player
            opp = 1 - p
            owe = street_invest[opp] - street_invest[p]

            if a == 'f':
                # Fold (only meaningful when facing a bet)
                folded[p] = True
                terminal = True
                break
            elif a == 'c':
                # Check (owe == 0)
                if last_aggressor == -1:
                    last_aggressor = -2
                    current_player = opp
                else:
                    advance_street()
            elif a == 'k':
                # Call — always ends the street (advance immediately)
                call_amt = min(owe, stacks[p])
                stacks[p] -= call_amt
                street_invest[p] += call_amt
                pot += call_amt
                advance_street()
            elif a == 'r':
                raise_add = bet_amount()
                total = owe + raise_add
                total = min(total, stacks[p])
                stacks[p] -= total
                street_invest[p] += total
                pot += total
                raises_this_street += 1
                last_aggressor = p
                current_player = opp
            elif a == 'a':
                allin = stacks[p]
                stacks[p] = 0.0
                street_invest[p] += allin
                pot += allin
                all_in = True
                raises_this_street += 1
                last_aggressor = p
                current_player = opp

        p = current_player
        opp = 1 - p
        to_call = street_invest[opp] - street_invest[p]
        invested = [self.starting_stack - stacks[0],
                    self.starting_stack - stacks[1]]

        return {
            "stacks": stacks,
            "invested": invested,
            "street_invest": street_invest,
            "pot": pot,
            "to_call": max(0.0, to_call),
            "street_idx": street,
            "street_name": ["preflop", "flop", "turn", "river"][min(street, 3)],
            "raises_this_street": raises_this_street,
            "current_player": current_player,
            "last_aggressor": last_aggressor,
            "folded": folded,
            "all_in": all_in,
            "terminal": terminal,
        }

    def _current_street(self, history: History) -> str:
        return self._parse_state(history)["street_name"]

    def _visible_board(self, history: History) -> tuple:
        """Board cards visible at the current street."""
        full_board = history[2]
        street = self._parse_state(history)["street_idx"]
        n = [0, 3, 4, 5][min(street, 3)]
        return full_board[:n]

    def is_terminal(self, history: History) -> bool:
        actions = self._get_actions(history)
        if not actions:
            return False
        return self._parse_state(history)["terminal"]

    def terminal_payoffs(self, history: History) -> tuple[float, ...]:
        if not self.is_terminal(history):
            raise ValueError(f"Non-terminal: {history}")

        state = self._parse_state(history)
        invested = state["invested"]

        # Fold: folder loses their total investment, winner gains it.
        if state["folded"][0] or state["folded"][1]:
            folder = 0 if state["folded"][0] else 1
            winner = 1 - folder
            payoffs = [0.0, 0.0]
            payoffs[folder] = -invested[folder]
            payoffs[winner] = invested[folder]
            return tuple(payoffs)

        # Showdown — compare full 7-card hands, winner takes the loser's stake.
        p0_cards = history[0]
        p1_cards = history[1]
        board = history[2][:5]
        h0 = evaluate_7card(p0_cards + board)
        h1 = evaluate_7card(p1_cards + board)
        if h0 > h1:
            return (float(invested[1]), float(-invested[1]))
        elif h1 > h0:
            return (float(-invested[0]), float(invested[0]))
        else:
            return (0.0, 0.0)  # Tie

    def current_player(self, history: History) -> int:
        return self._parse_state(history)["current_player"]

    def info_set_key(self, history: History, player: int) -> InfoSetKey:
        """
        For Deep CFR this isn't used for table lookup —
        the neural network generalizes across info sets.
        But we implement it for compatibility.
        """
        my_cards = history[player]
        board = self._visible_board(history)
        actions = "".join(str(a) for a in self._get_actions(history))
        return f"{my_cards}|{board}|{actions}"

    def legal_actions(self, history: History) -> list[Action]:
        state = self._parse_state(history)
        if state["terminal"]:
            return []

        p = state["current_player"]
        owe = state["to_call"]
        bet = owe > 0.001
        can_raise = state["raises_this_street"] < self.max_raises_per_street
        my_stack = state["stacks"][p]
        pot = state["pot"]

        def bet_amount():
            effective_pot = pot + owe
            raise_add = effective_pot * self.raise_fractions[0]
            return min(raise_add, my_stack - owe)

        result = []
        if bet:
            result.append("f")  # fold
            result.append("k")  # call
        else:
            result.append("c")  # check

        if can_raise and my_stack > owe + 0.01:
            if bet_amount() > 0.01:
                result.append("r")

        # All-in: available if it differs from the (capped) raise, or no raise.
        if my_stack > owe + 0.01:
            allin_add = my_stack - owe
            raise_add = bet_amount() if (can_raise and my_stack > owe + 0.01) else -1.0
            if (not can_raise) or allin_add > raise_add + 0.01:
                result.append("a")

        return result

    def apply_action(self, history: History, action: Action) -> History:
        return history + (action,)

File: web/src/lib/cfr/postflop-nlhe.ts
/**
 * TypeScript port of PostflopNLHE (src/games/postflop_nlhe.py).
 *
 * Bit-identical state machine to the C++ NLHEGame::apply_action that trained
 * the deployed blueprint. The action alphabet matches both: 'f' fold, 'c' check,
 * 'k' call, 'r' raise, 'a' all-in. Cards are 0-51 (rank*4 + suit).
 *
 * Implements the ExtensiveFormGame interface from ./types so subgame solvers
 * can use it as their `baseGame`. Browser-native, no I/O.
 */

import type { ExtensiveFormGame, Action, History, InfoSetKey, Player } from './types'

/** History tuple as a TypeScript-friendly readonly tuple. */
export type NLHEHistory = readonly [
  readonly [number, number],          // p0 hole cards
  readonly [number, number],          // p1 hole cards
  readonly number[],                  // board cards (0-5)
  ...string[],                        // action chars: f, c, k, r, a
]

/** Replayed game state at the end of a history's actions. */
export interface NLHEState {
  stacks: [number, number]
  invested: [number, number]      // cumulative across hand
  streetInvest: [number, number]  // within this street
  pot: number
  toCall: number
  streetIdx: number               // 0=preflop, 1=flop, 2=turn, 3=river, 4=terminal
  streetName: 'preflop' | 'flop' | 'turn' | 'river'
  raisesThisStreet: number
  currentPlayer: 0 | 1
  lastAggressor: number           // -1 no bet, -2 one check done
  folded: [boolean, boolean]
  allIn: boolean
  terminal: boolean
}

export interface PostflopNLHEConfig {
  startingStack?:        number   // default 50 (chips, BB=2, SB=1)
  maxRaisesPerStreet?:   number   // default 1
  sb?:                   number   // default 1
  bb?:                   number   // default 2
  raiseFraction?:        number   // default 0.75 (pot fraction)
}

/** Externally provided hand evaluator. Returns higher = better hand. */
export type HandEvaluator = (cards: readonly number[]) => number

export class PostflopNLHE implements ExtensiveFormGame {
  readonly startingStack:      number
  readonly maxRaisesPerStreet: number
  readonly sb:                 number
  readonly bb:                 number
  readonly raiseFraction:      number
  /** Used for showdown comparison. Inject from caller (encoder.ts evaluator). */
  private readonly evaluator: HandEvaluator

  constructor(cfg: PostflopNLHEConfig & { evaluator: HandEvaluator }) {
    this.startingStack      = cfg.startingStack      ?? 50
    this.maxRaisesPerStreet = cfg.maxRaisesPerStreet ?? 1
    this.sb                 = cfg.sb                 ?? 1
    this.bb                 = cfg.bb                 ?? 2
    this.raiseFraction      = cfg.raiseFraction      ?? 0.75
    this.evaluator          = cfg.evaluator
  }

  numPlayers(): number { return 2 }

  initialHistories(): Array<[History, number]> {
    // PostflopNLHE is too large for tabular CFR — kept here for interface
    // compatibility. Subgame wrappers override initialHistories() with
    // hero_range × opp_range enumeration.
    throw new Error('PostflopNLHE is too large for tabular CFR. Wrap in a SubgameGame.')
  }

  /** Strip actions out of a history. */
  private getActions(history: History): string[] {
    const h = history as NLHEHistory
    return h.slice(3) as string[]
  }

  /**
   * Deterministic replay of the state machine — mirrors Python _parse_state
   * and C++ NLHEGame::apply_action exactly. CALL ends the street, CHECK uses
   * last_aggressor double-tap, raise sizing = (pot+owe) * raiseFraction.
   */
  parseState(history: History): NLHEState {
    const actions = this.getActions(history)

    const stacks: [number, number] = [this.startingStack - this.sb, this.startingStack - this.bb]
    let streetInvest: [number, number] = [this.sb, this.bb]
    let pot = this.sb + this.bb
    let street = 0
    let currentPlayer: 0 | 1 = 0       // SB acts first preflop
    let raisesThisStreet = 0
    let lastAggressor = -1
    const folded: [boolean, boolean] = [false, false]
    let allIn = false
    let terminal = false

    const advanceStreet = (): void => {
      street++
      raisesThisStreet = 0
      lastAggressor = -1
      streetInvest = [0, 0]
      if (street >= 4) {
        terminal = true
      } else {
        currentPlayer = 1
      }
    }

    const betAmount = (p: 0 | 1): number => {
      const owe = streetInvest[1 - p] - streetInvest[p]
      const effectivePot = pot + owe
      const raiseAdd = effectivePot * this.raiseFraction
      return Math.min(raiseAdd, stacks[p] - owe)
    }

    for (const a of actions) {
      if (terminal) break
      const p   = currentPlayer
      const opp: 0 | 1 = (1 - p) as 0 | 1
      const owe = streetInvest[opp] - streetInvest[p]

      if (a === 'f') {
        folded[p] = true
        terminal = true
        break
      } else if (a === 'c') {
        if (lastAggressor === -1) {
          lastAggressor = -2
          currentPlayer = opp
        } else {
          advanceStreet()
        }
      } else if (a === 'k') {
        const callAmt = Math.min(owe, stacks[p])
        stacks[p] -= callAmt
        streetInvest[p] += callAmt
        pot += callAmt
        advanceStreet()
      } else if (a === 'r') {
        const raiseAdd = betAmount(p)
        let total = owe + raiseAdd
        total = Math.min(total, stacks[p])
        stacks[p] -= total
        streetInvest[p] += total
        pot += total
        raisesThisStreet++
        lastAggressor = p
        currentPlayer = opp
      } else if (a === 'a') {
        const allinAdd = stacks[p]
        stacks[p] = 0
        streetInvest[p] += allinAdd
        pot += allinAdd
        allIn = true
        raisesThisStreet++
        lastAggressor = p
        currentPlayer = opp
      }
    }

    const p   = currentPlayer
    const opp: 0 | 1 = (1 - p) as 0 | 1
    const toCall = Math.max(0, streetInvest[opp] - streetInvest[p])
    const invested: [number, number] = [
      this.startingStack - stacks[0],
      this.startingStack - stacks[1],
    ]
    const streetNames = ['preflop', 'flop', 'turn', 'river'] as const

    return {
      stacks, invested, streetInvest, pot, toCall,
      streetIdx: street,
      streetName: streetNames[Math.min(street, 3)],
      raisesThisStreet, currentPlayer, lastAggressor,
      folded, allIn, terminal,
    }
  }

  /** Visible board cards given the current street. */
  visibleBoard(history: History): readonly number[] {
    const h = history as NLHEHistory
    const fullBoard = h[2]
    const street = this.parseState(history).streetIdx
    const n = [0, 3, 4, 5][Math.min(street, 3)]
    return fullBoard.slice(0, n)
  }

  isTerminal(history: History): boolean {
    const actions = this.getActions(history)
    if (actions.length === 0) return false
    return this.parseState(history).terminal
  }

  terminalPayoffs(history: History): number[] {
    if (!this.isTerminal(history)) throw new Error('Non-terminal history')
    const state = this.parseState(history)
    const invested = state.invested

    // Fold path
    if (state.folded[0] || state.folded[1]) {
      const folder = state.folded[0] ? 0 : 1
      const winner = 1 - folder
      const out = [0, 0]
      out[folder] = -invested[folder]
      out[winner] =  invested[folder]
      return out
    }

    // Showdown
    const h = history as NLHEHistory
    const p0Cards = h[0]
    const p1Cards = h[1]
    const board   = h[2].slice(0, 5)
    const h0 = this.evaluator([...p0Cards, ...board])
    const h1 = this.evaluator([...p1Cards, ...board])
    if (h0 > h1) return [ invested[1], -invested[1]]
    if (h1 > h0) return [-invested[0],  invested[0]]
    return [0, 0]  // tie
  }

  currentPlayer(history: History): Player {
    return this.parseState(history).currentPlayer
  }

  legalActions(history: History): Action[] {
    const state = this.parseState(history)
    if (state.terminal) return []
    const p = state.currentPlayer
    const owe = state.toCall
    const bet = owe > 0.001
    const canRaise = state.raisesThisStreet < this.maxRaisesPerStreet
    const myStack = state.stacks[p]
    const pot = state.pot

    const betAmount = (): number => {
      const effectivePot = pot + owe
      const raiseAdd = effectivePot * this.raiseFraction
      return Math.min(raiseAdd, myStack - owe)
    }

    // Actions encoded as integer ids matching the C++ enum
    //   0 = FOLD_OR_CHECK ('f' if facing bet, 'c' otherwise)
    //   1 = CALL ('k')
    //   2 = RAISE ('r')
    //   3 = ALL_IN ('a')
    const out: Action[] = []
    if (bet) {
      out.push(0)  // fold
      out.push(1)  // call
    } else {
      out.push(0)  // check
    }

    if (canRaise && myStack > owe + 0.01) {
      if (betAmount() > 0.01) out.push(2)
    }

    if (myStack > owe + 0.01) {
      const allinAdd = myStack - owe
      const raiseAdd = (canRaise && myStack > owe + 0.01) ? betAmount() : -1
      if (!canRaise || allinAdd > raiseAdd + 0.01) out.push(3)
    }

    return out
  }

  /** Map numeric action → char used in history strings. */
  private actionToChar(action: Action, facingBet: boolean): string {
    if (action === 0) return facingBet ? 'f' : 'c'
    if (action === 1) return 'k'
    if (action === 2) return 'r'
    if (action === 3) return 'a'
    throw new Error(`Unknown action: ${action}`)
  }

  applyAction(history: History, action: Action): History {
    const h = history as NLHEHistory
    const state = this.parseState(history)
    const facingBet = state.toCall > 0.001
    const ch = this.actionToChar(action, facingBet)
    return [...h, ch] as unknown as History
  }

  infoSetKey(history: History, player: Player): InfoSetKey {
    const h = history as NLHEHistory
    const myCards = h[player] as readonly [number, number]
    const board   = this.visibleBoard(history)
    const actions = this.getActions(history).join('')
    return `${myCards.join(',')}|${board.join(',')}|${actions}`
  }
}
